File: Problems/2140. (M) Solving Questions With Brainpower/Python3/solution.py
class Solution:
  def mostPoints(self, questions: List[List[int]]) -> int:
    # bottom-up dynamic programming (tabulation)

    n = len(questions)
    # dp[i] is maximum score solving problems in questions[i:]
    dp = [0] * n
    dp[-1] = questions[-1][0]

    for i in range(n-2, -1, -1):
      dp[i], c_s = questions[i]
      if i + c_s + 1 < n:
        dp[i] += dp[i + c_s + 1]
      dp[i] = max(dp[i], dp[i+1])

    return dp[0]

File: Problems/2140. (M) Solving Questions With Brainpower/Java/solution.java
class Solution {
  public long mostPoints(int[][] questions) {
    // bottom-up dynamic programming (tabulation)

    int n = questions.length;
    // dp[i] is maximum score solving problems in questions[i:]
    long[] dp = new long[n];
    dp[n-1] = questions[n-1][0];

    for (int i = n-2; i >= 0; i--) {
      dp[i] = questions[i][0];
      int c_s = questions[i][1];
      if (i + c_s + 1 < n)
        dp[i] += dp[i + c_s + 1];
      dp[i] = Math.max(dp[i], dp[i+1]);
    }
    return dp[0];
  }
}
